### PR TITLE
sr_coex: Add support for Wi-Fi and BLE co-existence

### PR DESCRIPTION
**Pull request by muraliThokala**
This is for Wi-Fi and Short Range (SR) coexistence sample that demonstrates coexistence between WLAN and SR radios such as Bluetooth Low Energy (BLE) and Thread in 2.4GHz frequency. 

### Summary changes

This sample needed changes in the following samples/modules.
1. BLE throughput sample - By default BLE throughput is a file based one. To have BLE throughput like WLAN time-based throughput (so that both WLAN and BLE throughput runs simultaneously for the same duration) updates are done to BLE throughput sample to have support for both file and time-based approaches.
2. RPU - RPU updates are available as a firmware patch.
3. Wi-Fi Driver - Updates done to support multiple commands refactor the SR coexistence command in the
    Driver-RPU interface. 
4.  zperf - Updates done to have non-blocking zperf application.

### CODE REVIEW

File: samples/wifi/sr_coex/README.rst [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+| throughput    |              | Program coexistence sample application on nRF7002 DK.  |
++---------------+--------------+--------------------------------------------------------+
+| BLE only      | N.A          | Program BLE throughput application on nRF5340 DK and   |
+| throughput    |              | Select role as peripheral.                             |
+|               |              | Program coexistence sample application on nRF7002 DK.  |
++---------------+--------------+--------------------------------------------------------+
+| Wi-Fi and BLE | Disabled/    | Run Wi-Fi iperf in server mode on test PC.             |
+| combined      | Enabled      | Program BLE throughput application on nRF5340 DK and   |
+| throughput    |              | Select role as peripheral.                             |
+|               |              | Program coexistence sample application on nRF7002 DK.  |
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:186]:
```suggestion
| Wi-Fi only    | NA           | Run Wi-Fi **iperf** in server mode on test PC.         |
| throughput    |              | Program coexistence sample application on nRF7002 DK.  |
+---------------+--------------+--------------------------------------------------------+
| Bluetooth LE only      | NA          | Program BLE throughput application on nRF5340 DK and   |
| throughput    |              | select role as peripheral.                             |
|               |              | Program coexistence sample application on nRF7002 DK.  |
+---------------+--------------+--------------------------------------------------------+
| Wi-Fi and BLE | Disabled/    | Run Wi-Fi **iperf** in server mode on test PC.             |
| combined      | Enabled      | Program Bluetooth LE throughput application on nRF5340 DK and   |
| throughput    |              | select role as peripheral.                             |
|               |              | Program coexistence sample application on nRF7002 DK.  |
```
Again, messed up the table, I'm afraid. Just focused on the text. Bear with me.

File: samples/wifi/sr_coex/README.rst [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
++---------------+--------------+--------------------------------------------------------+
+
+Wi-Fi throughput result appears on test PC terminal on which iperf server is run. BLE throughput result appears on minicom terminal connected to nRF5340 DK.
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:189]:
```suggestion
The Wi-Fi throughput result appears on the test PC terminal on which **iperf** server is run.
The Bluetooth LE throughput result appears on the minicom terminal connected to the nRF5340 DK.
```

File: samples/wifi/sr_coex/README.rst [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+
+Results
+*******
+
+Summary of results obtained when coexistence tests are run for different Wi-Fi operating bands and antenna configurations with Wi-Fi mode set to 802.11n are given below. These tests are run keeping test setup inside an RF shield box. So, these results are representative and might change with changes in test setup.
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:194]:
```suggestion
The following table collects a summary of results obtained when coexistence tests are run for different Wi-Fi operating bands and antenna configurations with Wi-Fi mode set to 802.11n.
The tests are run with the test setup inside an RF shield box.
Therefore, the results are representative and might change with adjustments in the test setup.
```

File: samples/wifi/sr_coex/README.rst [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+
+**Wi-Fi in 2.4GHz**
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:197]:
```suggestion
Wi-Fi in 2.4 GHz
================
```

File: samples/wifi/sr_coex/README.rst [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+*Separate antennas*
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:199]:
```suggestion
Separate antennas
```

File: samples/wifi/sr_coex/README.rst [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:200]: BLE is a banned abbreviation. Change it to Bluetooth LE in all tables.

File: samples/wifi/sr_coex/README.rst [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+| central              |                    |                 |
++----------------------+--------------------+-----------------+
+| Wi-Fi and BLE,       | 9.9                | 145             |
+| coexistence disabled |                    |                 |
++----------------------+--------------------+-----------------+
+| Wi-Fi and BLE,       | 8.3                | 478             |
+| coexistence enabled  |                    |                 |
++----------------------+--------------------+-----------------+
+
+*Shared antenna*
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:218]:
```suggestion
Shared antenna
```

File: samples/wifi/sr_coex/README.rst [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+| central              |                    |                 |
++----------------------+--------------------+-----------------+
+| Wi-Fi and BLE,       | 10.2               | 29              |
+| coexistence disabled |                    |                 |
++----------------------+--------------------+-----------------+
+| Wi-Fi and BLE,       | 6.2                | 749             |
+| coexistence enabled  |                    |                 |
++----------------------+--------------------+-----------------+
+
+**Wi-Fi in 5GHz**
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:237]:
```suggestion
Wi-Fi in 5 GHz
=============
```

File: samples/wifi/sr_coex/README.rst [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+*Separate antennas*
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:239]:
```suggestion
Separate antennas
```

File: samples/wifi/sr_coex/README.rst [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+| central              |                    |                 |
++----------------------+--------------------+-----------------+
+| Wi-Fi and BLE,       | 10.2               | 1188            |
+| coexistence disabled |                    |                 |
++----------------------+--------------------+-----------------+
+| Wi-Fi and BLE,       | 10.2               | 1208            |
+| coexistence enabled  |                    |                 |
++----------------------+--------------------+-----------------+
+
+*Shared antenna*
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:258]:
```suggestion
Shared antenna
```

File: samples/wifi/sr_coex/README.rst [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+| coexistence disabled |                    |                 |
++----------------------+--------------------+-----------------+
+| Wi-Fi and BLE,       | 10.2               | 1191            |
+| coexistence enabled  |                    |                 |
++----------------------+--------------------+-----------------+
+
+Sample output
+*************
+
+Screenshots of coexistence test results obtained for separate antenna configuration with Wi-Fi mode set to 802.11n are given below. These tests are run when WLAN connected to an AP in 2.4GHz band. In these, left-side result shows Wi-Fi throughput that appears on test PC terminal in which Wi-Fi iperf server is run and right-side result shows BLE throughput that appears on minicom terminal in which BLE throughput sample is run.
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:280]:
```suggestion
The following screenshots show coexistence test results obtained for separate antenna configuration with Wi-Fi mode set to 802.11n.
These tests were run with WLAN connected to an AP in 2.4 GHz band.
In the images, left-side result shows Wi-Fi throughput that appears on a test PC terminal in which Wi-Fi **iperf** server is run and right-side result shows Bluetooth LE throughput that appears on a minicom terminal in which the Bluetooth LE throughput sample is run.
```

File: samples/wifi/sr_coex/README.rst [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+
+.. figure:: images/wlan.jpg
+     :width: 711px
+     :align: center
+     :alt: Wi-Fi only throughput
+
+     Wi-Fi only throughput 10.2Mbps
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:287]:
```suggestion
     Wi-Fi only throughput 10.2 Mbps
```

File: samples/wifi/sr_coex/README.rst [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+
+.. figure:: images/ble.jpg
+     :width: 711px
+     :align: center
+     :alt: BLE only throughput
+
+     BLE only throughput: 1107kbps
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:294]:
```suggestion
     Bluetooth LE only throughput: 1107 kbps
```

File: samples/wifi/sr_coex/README.rst [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+
+.. figure:: images/wlan_ble_cd.jpg
+     :width: 711px
+     :align: center
+     :alt: Wi-Fi and BLE CD
+
+     Wi-Fi and BLE throughput, coexistence disabled: Wi-Fi 9.9Mbps and BLE 145kbps
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:301]:
```suggestion
     Wi-Fi and Bluetooth LE throughput, coexistence disabled: Wi-Fi 9.9 Mbps and BLE 145 kbps
```

File: samples/wifi/sr_coex/README.rst [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+
+.. figure:: images/wlan_ble_cd.jpg
+     :width: 711px
+     :align: center
+     :alt: Wi-Fi and BLE CE
+
+     Wi-Fi and BLE throughput, coexistence enabled: Wi-Fi 8.3Mbps and BLE 478kbps
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:308]:
```suggestion
     Wi-Fi and Bluetooth LE throughput, coexistence enabled: Wi-Fi 8.3 Mbps and Bluetooth LE 478 kbps
```

File: samples/wifi/sr_coex/README.rst [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:309]: Remove extra empty lines. One is enough.

File: samples/wifi/sr_coex/README.rst [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+************
+
+This sample uses the following `sdk-nrfxlib`_ library:
+
+* :ref:`nrfxlib:nrf_security`
+
+
+.. _Threads: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/kernel/services/threads/index.html#threads
+
+.. _Network Traffic Generator: Network Traffic Generator — Zephyr Project Documentation
```
Comment (peknis) [inline samples/wifi/sr_coex/README.rst:323]: Instead of introducing the link targets here, insert them in doc/nrf/links.txt

**Review by peknis — CHANGES_REQUESTED** [review]

File: samples/wifi/sr_coex/README.rst [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+
+     Wi-Fi coexistence reference test and evaluation setup
+
+The following table provides more details on the sample or application that runs on DUT and peer devices.
+This setup is kept in a shielded test enclosure box (for example a Ramsey box).
+
++--------------+----------------+------------------------------------------------------------------------------+
+| Device       | Application    |                             Details                                          |
++==============+================+==============================================================================+
+| nRF7002 DK   | coex sample    | The sample runs Wi-Fi throughputs, Bluetooth LE throughputs                  |
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:48]: Bluetooth LE coexistence sample

File: samples/wifi/sr_coex/README.rst [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+
+
+Wi-Fi in 2.4 GHz
+================
+
+Separate antennas
+
++------------------------+--------------------+-------------------+
+| Test case              | Wi-Fi UDP Tx       | Bluetooth LE      |
+|                        | throughput in Mbps | throughputin kbps |
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:239]:
```suggestion
|                        | throughput in Mbps | throughput in kbps |
```

File: samples/wifi/sr_coex/README.rst [hunk 20]
```diff
@@ -0,0 +1,359 @@
...
++------------------------+--------------------+-------------------+
+| Wi-Fi and Bluetooth LE,| 8.3                | 478               |
+| coexistence enabled    |                    |                   |
++------------------------+--------------------+-------------------+
+
+Shared antenna
+
++------------------------+--------------------+-------------------+
+| Test case              | Wi-Fi UDP Tx       | Bluetooth LE      |
+|                        | throughput in Mbps | throughputin kbps |
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:258]:
```suggestion
|                        | throughput in Mbps | throughput in kbps |
```

File: samples/wifi/sr_coex/README.rst [hunk 21]
```diff
@@ -0,0 +1,359 @@
...
++------------------------+--------------------+-------------------+
+
+Wi-Fi in 5 GHz
+=============
+
+Separate antennas
+
++------------------------+--------------------+-------------------+
+| Test case              | Wi-Fi UDP Tx       | Bluetooth LE      |
+|                        | throughput in Mbps | throughputin kbps |
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:280]:
```suggestion
|                        | throughput in Mbps | throughput in kbps |
```

File: samples/wifi/sr_coex/README.rst [hunk 22]
```diff
@@ -0,0 +1,359 @@
...
++------------------------+--------------------+-------------------+
+| Wi-Fi and Bluetooth LE,| 10.2               | 1208              |
+| coexistence enabled    |                    |                   |
++------------------------+--------------------+-------------------+
+
+Shared antenna
+
++------------------------+--------------------+-------------------+
+| Test case              | Wi-Fi UDP Tx       | Bluetooth LE      |
+|                        | throughput in Mbps | throughputin kbps |
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:299]:
```suggestion
|                        | throughput in Mbps | throughput in kbps |
```

File: samples/wifi/sr_coex/README.rst [hunk 23]
```diff
@@ -0,0 +1,359 @@
...
+
+  * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_ONLY` for Wi-Fi only test
+  * :kconfig:option:`CONFIG_TEST_TYPE_BLE_ONLY` for Bluetooth LE only test
+  * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_BLE` for concurrent Wi-Fi and Bluetooth LE test.
+
+  Based on the required test, set only one of these to ``y`` while the other two have to be set to ``n``.
+* Test duration: Use the :kconfig:option:`CONFIG_WIFI_TEST_DURATION` Kconfig option to set the duration of the Wi-Fi test and :kconfig:option:`CONFIG_BLE_TEST_DURATION` for the Bluetooth LE test.
+  The units are in milliseconds.
+  For example, to set the tests for 20 seconds, set the respective values to ``20000``.
+  For the concurrent Wi-Fi and Bluetooth LE test, make sure that both of these are set to the same duration to ensure maximum overlap.
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:97]:
```suggestion
  For the concurrent Wi-Fi and Bluetooth LE test, make sure that both are set to the same duration to ensure maximum overlap.
```

File: samples/wifi/sr_coex/README.rst [hunk 24]
```diff
@@ -0,0 +1,359 @@
...
+.. note::
+   Use the same test duration value for :kconfig:option:`CONFIG_WIFI_TEST_DURATION`, :kconfig:option:`CONFIG_BLE_TEST_DURATION` and ``BT_THROUGHPUT_DURATION``.
+
+Connecting to and programming DKs
+=================================
+
+Connecting to DKs
+*****************
+
+After the DKs are connected to the test PC through USB connectors and powered on, open a suitable terminal and run the following command:
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:153]:
```suggestion
After the DKs are connected to the test PC through USB connectors and powered on, open a suitable terminal, and run the following command:
```

File: samples/wifi/sr_coex/README.rst [hunk 25]
```diff
@@ -0,0 +1,359 @@
...
+* File or time-based throughput: Use ``BT_THROUGHPUT_FILE`` to select file or time-based throughput test.
+  Set it to ``n`` to enable time-based throughput test only when running Bluetooth LE throughput in central role.
+* Test duration: Use ``BT_THROUGHPUT_DURATION`` to set the duration of the Bluetooth LE throughput test only when running Bluetooth LE throughput in central role.
+  The units are in milliseconds.
+
+.. note::
+   Use the same test duration value for :kconfig:option:`CONFIG_WIFI_TEST_DURATION`, :kconfig:option:`CONFIG_BLE_TEST_DURATION` and ``BT_THROUGHPUT_DURATION``.
+
+Connecting to and programming DKs
+=================================
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:148]: Could be removed as the subheadings are there.

File: samples/wifi/sr_coex/README.rst [hunk 26]
```diff
@@ -0,0 +1,359 @@
...
+
+Connecting to DKs
+*****************
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:151]:
```suggestion
Connecting to DKs
=================
```

File: samples/wifi/sr_coex/README.rst [hunk 27]
```diff
@@ -0,0 +1,359 @@
...
+In the example, use ttyACM1 to connect to the board with device ID ``1050043161``.
+Similarly, use ttyACM3 connect to the board with device ID ``1050724225``.
+
+.. code-block:: console
+
+   $ minicom -D /dev/ttyACM1 -b 115200
+   $ minicom -D /dev/ttyACM3 -b 115200
+
+Programming DKs
+***************
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:176]:
```suggestion
Programming DKs
===============
```

**Review by umapraseeda — CHANGES_REQUESTED** [review]

File: samples/wifi/sr_coex/README.rst [hunk 28]
```diff
@@ -0,0 +1,388 @@
...
+
+* Antenna configuration: Use the :kconfig:option:`CONFIG_COEX_SEP_ANTENNAS` Kconfig option to select the antenna configuration.
+  Set it to ``y`` to enable separate antennas and ``n`` to enable shared antenna.
+* Test modes: Use the following Kconfig options to select the required test case:
+
+  * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_ONLY` for Wi-Fi only test
+  * :kconfig:option:`CONFIG_TEST_TYPE_BLE_ONLY` for Bluetooth LE only test
+  * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_BLE` for concurrent Wi-Fi and Bluetooth LE test.
+
+  Based on the required test, set only one of these to ``y`.
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:86]:
```suggestion
  * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_BLE` for concurrent Wi-Fi and Bluetooth LE test.

    Based on the required test, set only one of these to ``y`.
```

File: samples/wifi/sr_coex/README.rst [hunk 29]
```diff
@@ -0,0 +1,388 @@
...
+
+   west build -b nrf7002dk_nrf5340_cpuapp -- -DCONFIG_MPSL_CX=n -Dhci_rpmsg_CONFIG_MPSL_CX=n
+
+* Build with coexistence enabled:
+
+.. code-block:: console
+
+   west build -b nrf7002dk_nrf5340_cpuapp -- -DCONFIG_MPSL_CX=y -Dhci_rpmsg_CONFIG_MPSL_CX=y
+
+The generated hex file to be used is :file:`build/zephyr/merged_domains.hex`.
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:121]:
```suggestion
The generated HEX file to be used is :file:`build/zephyr/merged_domains.hex`.
```

**Review by krish2718 — APPROVED** [review]

**Review by bama-nordic — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_coex.h [hunk 30]
```diff
@@ -0,0 +1,73 @@
...
+ *             Returns zero upon successful configuration.
+ *             Returns non-zero upon unsuccessful configuration.
+ */
+int nrf_wifi_coex_config_pta(uint32_t wlan_band);
+
+
+/**
+ * @function   nrf_wifi_config_sr_switch()
+ *
+ * @brief      Function used to configure SR side switch (nRF5340 side switch in nRF7002 DK).
```
Comment (bama-nordic) [inline drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_coex.h:39]: Explain that this API configures the SR side antenna switch as per the selection of shared/separate antenna mentioning CONFIG_COEX_SEP_ANTENNAS symbol needs to be set appropriately.

**Review by bama-nordic — CHANGES_REQUESTED** [review]

**Review by bama-nordic — APPROVED** [review]
LGTM

File: samples/wifi/sr_coex/README.rst [hunk 31]
```diff
@@ -0,0 +1,400 @@
...
+     :align: center
+     :alt: Wi-Fi and Bluetooth LE CD
+
+     Wi-Fi and Bluetooth LE throughput, coexistence disabled: Wi-Fi 9.9 Mbps and Bluetooth LE 145 kbps
+
+.. figure:: images/wlan_ble_ce.jpg
+     :width: 711px
+     :align: center
+     :alt: Wi-Fi and Bluetooth LE CE
+
```
Comment (umapraseeda) [inline samples/wifi/sr_coex/README.rst:382]: @richabp, FYI the images might need rework (as discussed offline) and might need to be moved into the common images folder.

**Comment by umapraseeda** [comment]
Please add an entry in the changelog: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/releases/release-notes-changelog.html#wi-fi-samples

**Review by sachinthegreen — APPROVED** [review]

File: doc/nrf/releases/release-notes-changelog.rst [hunk 32]
```diff
@@ -127,7 +127,9 @@ nRF IEEE 802.15.4 radio driver
 Wi-Fi
 -----
 
-|no_changes_yet_note|
+* Added:
+
+  * Sample demonstrating Wi-Fi Bluetooth LE coexistence
```
Comment (umapraseeda) [inline doc/nrf/releases/release-notes-changelog.rst:132]:
```suggestion
```

File: doc/nrf/releases/release-notes-changelog.rst [hunk 33]
```diff
@@ -350,8 +352,9 @@ Zigbee samples
 
 Wi-Fi samples
 -------------
+  * Added:
 
-|no_changes_yet_note|
+    * Wi-Fi Bluetooth LE coexistence sample
```
Comment (umapraseeda) [inline doc/nrf/releases/release-notes-changelog.rst:357]:
```suggestion
* Added:

  * :ref:`wifi_sr_coex_sample` sample demonstrating Wi-Fi Bluetooth LE coexistence.
 ```

**Review by umapraseeda — APPROVED** [review]

File: dts/bindings/radio_coex/nordic,nrf700x-coex.yaml [hunk 34]
```diff
@@ -34,8 +34,7 @@ properties:
         required: false
         description: |
             GPIO of the SOC controlling the Priority (STATUS1) pin (in 4-wire
-            coex case) of the nRF7002
-
+            coex case) of the nRF7002.
```
Comment (jciupis) [inline dts/bindings/radio_coex/nordic,nrf700x-coex.yaml:37]: Please restore the empty line between properties.

**Review by jciupis — APPROVED** [review]

**Review by peknis — APPROVED** [review]

**Pull request merged**